inspire_json_merger/postprocess.py

```diff
@@ -0,0 +1,133 @@
+    """Cleans up ordering information in merged record."""
+    if 'authors' in merged:
+        for author in merged["authors"]:
+            author.pop(ORDER_KEY, None)
+    return merged
+
+
+def postprocess_conflicts(conflicts, merged):
+    """Postprocessing conflicts to display only useful conflicts.
+
+    Before flattening and serializing to JSON patch, MERGE conflict looks like this:
+    ('MANUAL_MERGE', ('authors',), ( pmap(<ROOT>), pmap(<HEAD>), pmap(<UPDATE>) )),
+    This function takes out all MANUAL_MERGE and do as follows:
+        - Ignore ROOT
+        - Update merged result with HEAD from conflict by inserting the author
+            from HEAD at the right position.
+        - Look for ADD_BACK_TO_HEAD conflict with identical content as HEAD
+            change and remove it as it's already added to HEAD
+    Args:
+        conflicts(list): List of all possible conflicts
+        merged(dict): Merged document
+
+    Returns: A tuple containing the resulted merged record in json format and a
+        an list containing all generated conflicts.
+    """
+    new_conflicts = []
+    possible_duplicates = []
+    for conflict in conflicts:
+        conflict_type, conflict_location, conflict_content = conflict
+        if conflict_type == "MANUAL_MERGE" and conflict_location[0] == "authors":
+            new_conflict, merged, head = _process_author_manual_merge_conflict(conflict, merged)
+            if new_conflict:
+                new_conflicts.append(new_conflict)
+                possible_duplicates.append(head)
+        else:
+            new_conflicts.append(conflict)
+    new_conflicts = _remove_duplicates(new_conflicts, set(possible_duplicates))
+    return new_conflicts, merged
+
+
+def _remove_duplicates(conflicts, possible_duplicates):
+    new_conflicts = []
+    for conflict in conflicts:
+        conflict_type, conflict_location, conflict_content = conflict
+        if conflict_type == "ADD_BACK_TO_HEAD" and conflict_location[0] == "authors" and conflict_content in possible_duplicates:
+            continue
+        else:
+            new_conflicts.append(conflict)
+    return new_conflicts
+
+
+def _process_author_manual_merge_conflict(conflict, merged):
+    """Process author `MANUAL_MERGE` conflict
+
+    Conflict object is an tuple containing
+    (conflict_type, conflict_location, conflict_data)
+    where `conflict_data` is a tuple of: (ROOT, HEAD, UPDATE)
+    """
+    _, _, (root, head, update) = conflict
+    if head not in merged["authors"]:
+        position, merged['authors'] = _insert_author(dict(head), merged['authors'])
+        new_conflict = Conflict("SET_FIELD", ("authors/{position}".format(position=position),), update)
+        return new_conflict, merged, head
+    return None, merged, head
+
+
+def _insert_author(head, merged):
+    position = head[ORDER_KEY]
+    for idx, element in enumerate(merged):
+        if idx > position and (ORDER_KEY not in element or element[ORDER_KEY] > position):
```

Comment:
you didn't answer this, what is `idx > position` needed for?

Reply:
Discussed IRL